File: app/src/main/java/com/me/noteapp/data/helper/DBHelper.java
package com.me.noteapp.data.helper;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteException;
import android.database.sqlite.SQLiteOpenHelper;

import com.me.noteapp.config.DBConfig;
import com.me.noteapp.entity.History;
import com.me.noteapp.entity.Item;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

public class DBHelper extends SQLiteOpenHelper {
    private static DBHelper sInstance;
    private static DBConfig cfg = DBConfig.getInstance();
    private static final List<String> SQL_CREATE_TABLES = cfg.getCreateTablesQuery();
    private static final List<String> SQL_DELETE_TABLES = cfg.getDeleteTablesQuery();
    private static final String name = cfg.DB_NAME;
    private static int version = 1;
    private ContentValues values;

    public static synchronized DBHelper getInstance(Context context) {
        if (sInstance == null) {
            sInstance = new DBHelper(context.getApplicationContext());
        }
        return sInstance;
    }

    public synchronized boolean insertItem(Item item) {
        SQLiteDatabase db = getWritableDatabase();
        values = new ContentValues();
        values.put("dAt", item.getdAt().getTime());
        values.put("content", item.getContent());
        long res = 0;
        res = db.insert("items", null, values);
        return res != -1;
    }

    public boolean updateItem(Item item) {
        SQLiteDatabase db = getWritableDatabase();
        values = new ContentValues();
        String id = String.valueOf(item.getdAt().getTime());
        values.put("dAt", id);
        values.put("content", item.getContent());

        db.beginTransaction();
        long res;
        try {
            res = db.update("items", values, "dAt LIKE ?", new String[]{id});
            insertHistory(new History(item.getdAt().getTime(), new Date()));
            db.setTransactionSuccessful();
        } finally {
            db.endTransaction();
            db.close();
        }
        return res > 0;
    }

    public boolean deleteItem(String itemId) {
        SQLiteDatabase db = getWritableDatabase();
        long res = db.delete("items", "dAt LIKE ?", new String[]{itemId});
        return res > 0;
    }

    public Item read(String id) {
        SQLiteDatabase db = getWritableDatabase();
        Cursor cursor;
        cursor = db.rawQuery("select * from items where dAt = ?;", new String[]{id});
        if (cursor.moveToFirst()) {
            Date dAt = new Date(cursor.getLong(cursor.getColumnIndex("dAt")));
            String content = cursor.getString(cursor.getColumnIndex("content"));
            return new Item(dAt, content);
        }
        return null;
    }

    public boolean insertHistory(History h) {
        SQLiteDatabase db = getWritableDatabase();
        values = new ContentValues();
        values.put("itemId", h.getItemId());
        values.put("id", h.getId().getTime());
        long res = 0;
        res = db.insert("histories", null, values);
        return res != -1;
    }

    public List<History> getHistories(String id) {
        SQLiteDatabase db = getWritableDatabase();
        Cursor cursor;
        List<History> histories = new ArrayList<>();
        cursor = db.rawQuery("select * from histories where itemId = ?;", new String[]{id});
        Date hId;
        long itemId;
        if (cursor.moveToFirst()) {
            itemId = cursor.getLong(cursor.getColumnIndex("itemId"));
            while (cursor.isAfterLast() == false) {
                hId = new Date(cursor.getLong(cursor.getColumnIndex("id")));
                histories.add(new History(itemId, hId));
                cursor.moveToNext();
            }
        }
        return histories;
    }

    public List<Item> readAllItems() {
        SQLiteDatabase db = getWritableDatabase();
        List<Item> items = new ArrayList<>();
        Cursor cursor;
        try {
            cursor = db.rawQuery("select * from items i order by i.dAt desc;", null);
        } catch (SQLiteException e) {
            return new ArrayList();
        }

        Date dAt;
        String content;

        if (cursor.moveToFirst()) {
            while (cursor.isAfterLast() == false) {
                dAt = new Date(cursor.getLong(cursor.getColumnIndex("dAt")));
                content = cursor.getString(cursor.getColumnIndex("content"));
                items.add(new Item(dAt, content));
                cursor.moveToNext();
            }
        }
        return items;
    }

    private DBHelper(Context context) {
        super(context, name, null, version);
    }

    @Override
    public void onCreate(SQLiteDatabase db) {
        for (String sql : SQL_CREATE_TABLES) db.execSQL(sql);
    }

    @Override
    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
        for (String sql : SQL_DELETE_TABLES) db.execSQL(sql);
        onCreate(db);
    }

    @Override
    public void onDowngrade(SQLiteDatabase db, int oldVersion, int newVersion) {
        onUpgrade(db, oldVersion, newVersion);
    }
}

File: app/src/main/java/com/me/noteapp/data/helper/DBHelper.kt
package com.me.noteapp.data.helper

import android.content.ContentValues
import android.content.Context
import android.database.Cursor
import android.database.sqlite.SQLiteConstraintException
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteException
import android.database.sqlite.SQLiteOpenHelper
import com.me.noteapp.config.DBConfig
import com.me.noteapp.entity.Item
import java.util.*
import kotlin.collections.ArrayList

class DBHelper(ctx: Context) : SQLiteOpenHelper(ctx, DATABASE_NAME, null, DATABASE_VERSION) {

    override fun onCreate(db: SQLiteDatabase) {
        db.execSQL(SQL_CREATE_TABLES)
    }

    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
        db.execSQL(SQL_DELETE_TABLES)
        onCreate(db)
    }

    override fun onDowngrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
        onUpgrade(db, oldVersion, newVersion)
    }

    @Throws(SQLiteConstraintException::class)
    fun insertItem(item: Item): Boolean {
        val db = writableDatabase

        val values = ContentValues()
        values.put("dAt", item.getdAt().time)
        values.put("content", item.content)

        val res = db.insert("items", null, values)
        return !res.equals(-1)
    }

    fun updateItem(item: Item): Boolean {
        val db = writableDatabase

        val values = ContentValues()
        val id: Long = item.getdAt().time
        values.put("dAt", id)
        values.put("content", item.content)

        val res = db.update("items", values, "dAt LIKE ?", arrayOf(id.toString()))
        return res > 0
    }

    fun deleteItem(itemId: String): Boolean {
        val db = writableDatabase

        val res = db.delete("items", "dAt LIKE ?", arrayOf(itemId))
        return res > 0
    }

    fun read(id: String): Item? {
        val db = readableDatabase
        var cursor: Cursor? = null
        cursor = db.rawQuery("select * from items where dAt = ?;", Array<String>(1) { id })
        if (cursor!!.moveToFirst()) {
            val dAt = Date(cursor.getLong(cursor.getColumnIndex("dAt")))
            val content = cursor.getString(cursor.getColumnIndex("content"))
            return Item(dAt, content)
        }
        return null
    }

    fun readAllItems(): ArrayList<Item> {
        val items = ArrayList<Item>()
        val db = readableDatabase
        var cursor: Cursor? = null
        try {
            cursor = db.rawQuery("select * from items i order by i.dAt desc;", null)
        } catch (e: SQLiteException) {
            return ArrayList()
        }

        var dAt: Date
        var content: String

        if (cursor!!.moveToFirst()) {
            while (cursor.isAfterLast == false) {
                dAt = Date(cursor.getLong(cursor.getColumnIndex("dAt")))
                content = cursor.getString(cursor.getColumnIndex("content"))

                items.add(Item(dAt, content))
                cursor.moveToNext()
            }
        }
        return items
    }

    companion object {
        private var cfg: DBConfig = DBConfig.getInstance()

        fun createTablesQuery(): String {
            var SQL_CREATE_TABLES = ""
            cfg.tables.forEach {
                SQL_CREATE_TABLES += "CREATE TABLE ${it.first} (${it.second});"
            }
            return SQL_CREATE_TABLES
        }

        fun deleteTablesQuery(): String {
            var SQL_DELETE_TABLES = ""
            cfg.tables.forEach {
                SQL_DELETE_TABLES += "DROP TABLE IF EXISTS ${it.first};"
            }
            return SQL_DELETE_TABLES
        }


        val DATABASE_VERSION = 1
        val DATABASE_NAME = cfg.DB_NAME
        private val SQL_CREATE_TABLES = createTablesQuery()
        private val SQL_DELETE_TABLES = deleteTablesQuery()
    }
}